### PR TITLE
[Team Email Login] Admin enable/disable recaptcha during email signup

### PR DESCRIPTION
**Pull request by msarit**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [X] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description
***Update: I incorporated @lisasy's UX changes, and other requested QA changes to the PR**
This PR grants Forem admins the ability to require reCAPTCHA during email signup. The PR also implements the reCAPTCHA check during email signup.

## Related Tickets & Documents
None

## QA Instructions *(updated)*
1. Sign into your local app as an `admin` (or `super_admin`) with `single_resource_admin` role (important), navigate to `/admin/config` and toggle the Authentication section open.
2. Check the `Enable Google reCAPTCHA for email password registration` option, which should reveal the fields for the Google reCAPTCHA keys. Enter the keys (see guide right after this paragraph for how to obtain reCAPTCHA keys for `localhost`). Be sure to `Update Site Configuration` at the bottom of the section after making changes!
2. In an incognito window, navigate to `localhost:<port>/users/sign_up?state=email_signup` and test the signup with reCAPTCHA completed. You should be able to create an account successfully.
3. Run through the signup process again, but do not complete the reCAPTCHA. You should be blocked from registering a new account.

**Getting Google ReCaptcha Keys:**
1. Google "google reCAPTCHA admin" and navigate to your admin console
2. Click the "+" in the top right area to add keys for a new site
3. Enter any label
4. Choose `reCAPTCHA v2`
5. Add TWO domains - `example.com` (or something else random) and `localhost`. The first random domain prevents 
 errors being thrown on your local app, as technically Google does not support `localhost` as the sole domain.
6. Accept Terms and Submit

## Added tests?

- [X] yes
- [ ] no, because they aren't needed
- [ ] no, because I need help

## Added to documentation?

- [ ] docs.forem.com
- [ ] readme
- [X] no documentation needed

## [optional] Are there any post deployment tasks we need to perform?
None

## [optional] What gif best describes this PR or how it makes you feel?

woman blocked by guards

### CODE REVIEW

File: app/lib/constants/site_config.rb [hunk 1]
```diff
@@ -2,11 +2,15 @@ module Constants
   module SiteConfig
     DETAILS = {
       allow_email_password_registration: {
-        description: "Can users sign up with only email and password?",
+        description: "Can users sign up using their email and password?",
```
Comment (msarit) [inline app/lib/constants/site_config.rb:5]: I changed the language of the `description` here to something that makes more sense to me (and perhaps to @vaidehijoshi, as she mentioned being a little confused with the word `only` 😄 )

**Comment by joshpuetz** [comment]
Tagging team OSS for some auxiliary review!

File: app/views/shared/authentication/_email_registration_form.html.erb [hunk 2]
```diff
@@ -54,6 +59,14 @@
         <%= f.password_field :forem_owner_secret, autocomplete: "current-password", class: "crayons-textfield", required: true, placeholder: "As provided by your Forem host" %>
       </div>
     <% end %>
+
+    <% if SiteConfig.recaptcha_secret_key.present? &&
```
Comment (joshpuetz) [inline app/views/shared/authentication/_email_registration_form.html.erb:63]: **Suggestion**: this is a pretty beefy conditional in a view. Consider extracting this to a helper method.

File: spec/requests/registrations_spec.rb [hunk 3]
```diff
@@ -154,6 +175,34 @@
...
+      it "does not create user when valid params passed BUT recaptcha incomplete" do
+        post "/users", params:
+          { user: { name: "test #{rand(10)}",
+                    username: "haha_#{rand(10)}",
+                    email: "yoooo#{rand(100)}@yo.co",
+                    password: "PaSSw0rd_yo000",
+                    password_confirmation: "PaSSw0rd_yo000" } }
+        expect(User.all.size).to be 0
+      end
+    end
```
Comment (joshpuetz) [inline spec/requests/registrations_spec.rb:204]: **Suggestion** Test to make sure the appropriate flash/error message is displayed to the user.

**Review by joshpuetz — CHANGES_REQUESTED** [review]
Codewise added a few suggestions, but not blocking.

What is blocking however is the UI layout IMHO: it's a really odd flow to have the checkbox for enabling captcha in the Auth section of config, but then have to scroll down to a separate Recaptcha section to add keys. Looking for some input from @lisasy, but I'd suggest moving the recaptcha enable/disable box down to the Captcha section **or** moving the captcha key fields into the Authentication section.

**Comment by lisasy** [comment]
2020-10-15_Google-reCaptcha

@msarit I've gotten to look at this PR and think that there are some ways we can improve the UX and make it easier for someone to setup. I'm attaching a screenshot of what that solution could be. It would involve:

- Moving the current "Google reCAPTCHA" fields to be nested within the recaptcha toggle in [Admin]. We should only show these fields if the recaptcha toggle is enabled.
- Update the language for these labels so that there is more consistency here. I think we could also update its description, but I wanted to check in with you to better understand how people will retrieve their Google reCAPTCHA site and secret key. In this PR,  you wrote up instructions on how to get Google reCAPTCHA keys; are these directions only for people reviewing this PR, or is it also applicable to the admins who would be using this flow for their community? That could determine how we explain these instructions in-product.

**Comment by msarit** [comment]
Requesting re-review please! 🙏  @joshpuetz @lisasy @benhalpern

**Review by joshpuetz — APPROVED** [review]
Looks great, the new admin options layout is terrific!

**Review by benhalpern — APPROVED** [review]
This is a really well done PR.

File: app/views/shared/authentication/_email_registration_form.html.erb (hunk 2 above)
Comment (benhalpern) [reply app/views/shared/authentication/_email_registration_form.html.erb]: I was very pleased to see this when I showed up to review it. Very clean. 🤙

**Pull request merged**